In [1]:
f = open('day05_input.txt')
lines = f.readlines()
f.close()
tickets = [line[0:-1] for line in lines]

In [2]:
print(f'There are {len(tickets)} tickets. The first 5 are:')
tickets[0:5]

There are 817 tickets. The first 5 are:


['FBFFFFBRLL', 'FFBFFFFRLR', 'BFBBFFBRLL', 'FBBBBFFLLR', 'BFFFBFFLRL']

# Part 1

This is a 2D binary search problem. 'F'/'B' means front/back half (for rows), and 'R'/'L' means right/left half (for columns). There are 128 rows (0-127), and 8 columns (0-7).

The first 7 letters are either F or B, for rows, and the last three letters are either R or L, for columns.

In [3]:
ans = 0
for ticket in tickets:
    i, j = 0, 127
    for ch in ticket[0:7]:
        size = j - i + 1
        if ch == 'F':
            j -= size//2
        else:
            i += size//2
    m, n = 0, 7
    for ch in ticket[-3:]:
        size = n - m + 1
        if ch == 'L':
            n -= size//2
        else:
            m += size//2
    ans = max(ans, i*8 + m)
print(f'The max ID is {ans}.')

The max ID is 901.


# Part 2

Looks like one of the missing seats is yours. But it can't be at the very front or back, and the seats with +1 and -1 from yours should also be in the "tickets".

From the way the "ID" is calculated, the small IDs are the seats in the front, and the larger ones are in the back. First of all, we need to find out the missing seats:

In [4]:
all_seats = list(range(1,902))
for ticket in tickets:
    i, j = 0, 127
    for ch in ticket[0:7]:
        size = j - i + 1
        if ch == 'F':
            j -= size//2
        else:
            i += size//2
    m, n = 0, 7
    for ch in ticket[-3:]:
        size = n - m + 1
        if ch == 'L':
            n -= size//2
        else:
            m += size//2
    id = i*8 + m
    all_seats.remove(id)

print('The missing (or empty) seats are:')
print(all_seats)

The missing (or empty) seats are:
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 661]


By only looking at the list above, we know that seat **661** is the one.

To double check, we can remove all the consecutive seats from the list above:

In [5]:
previous_num = all_seats[0]
available_seats = set(all_seats)
ans = []
for num in all_seats[1:]:
    if num == previous_num + 1:
        available_seats.remove(num)
        if previous_num in available_seats:
            available_seats.remove(previous_num)
    previous_num = num
    
print(available_seats)

{661}
